game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java

```diff
@@ -0,0 +1,129 @@
+import games.strategy.triplea.delegate.battle.MustFightBattle.RetreatType;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+@AllArgsConstructor
+public class OffensiveGeneralRetreat implements BattleStep {
+
+  final BattleState battleState;
+
+  final BattleActions battleActions;
+
+  @Override
+  public List<String> getNames() {
+    if (canAttackerRetreat()
+        || canAttackerRetreatSeaPlanes()
+        || (battleState.isAmphibious()
+            && (canAttackerRetreatPartialAmphib() || canAttackerRetreatAmphibPlanes()))) {
+      return List.of(battleState.getAttacker().getName() + ATTACKER_WITHDRAW);
+    }
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.OFFENSIVE_GENERAL_RETREAT;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    retreatUnits(bridge);
+  }
+
+  public void retreatUnits(final IDelegateBridge bridge) {
+    if (battleState.isOver()) {
+      return;
+    }
+    final RetreatData retreatData;
+
+    if (battleState.isAmphibious()) {
+      retreatData = getAmphibiousRetreatData();
+      if (retreatData == null) {
+        return;
+      }
+
+    } else if (canAttackerRetreat()) {
+      retreatData =
+          RetreatData.of(RetreatType.DEFAULT, battleState.getAttackerRetreatTerritories());
+    } else {
+      return;
+    }
+
+    battleActions.queryRetreat(false, retreatData.retreatType, bridge, retreatData.retreatSites);
+  }
+
+  private @Nullable RetreatData getAmphibiousRetreatData() {
+    if (canAttackerRetreatPartialAmphib()) {
+      return RetreatData.of(
+          RetreatType.PARTIAL_AMPHIB, battleState.getAttackerRetreatTerritories());
+
+    } else if (canAttackerRetreatAmphibPlanes()) {
+      return RetreatData.of(RetreatType.PLANES, Set.of(battleState.getBattleSite()));
+
+    } else {
+      return null;
+    }
+  }
+
+  @Value(staticConstructor = "of")
+  private static class RetreatData {
+    RetreatType retreatType;
+    Collection<Territory> retreatSites;
+  }
+
+  private boolean canAttackerRetreat() {
+    return RetreatChecks.canAttackerRetreat(
+        battleState.getUnits(BattleState.Side.DEFENSE),
+        battleState.getGameData(),
+        battleState::getAttackerRetreatTerritories,
+        battleState.isAmphibious());
+  }
+
+  private boolean canAttackerRetreatSeaPlanes() {
+    return battleState.getBattleSite().isWater()
+        && battleState.getUnits(BattleState.Side.OFFENSE).stream().anyMatch(Matches.unitIsAir());
+  }
+
+  private boolean canAttackerRetreatPartialAmphib() {
+    if (Properties.getPartialAmphibiousRetreat(battleState.getGameData())) {
```

Comment:
Have you consider pushing this conditional to the callers? I think in both cases it would be the calling code easier to understand, it avoids having to jump around the class file to understand those locations. Furthermore, without this if statement the method body would be a simple stream & any match statement. Not that it matters too much, there would be a small performance boost as you wouldn't need to compute the full set of land units but could halt on the first match.

Reply:
I think it is better if the caller doesn't have to know what properties enable/disable partial amphibious retreat.

I've made the stream/anyMatch change, though.

Reply:
Hmm, it's a private method though.

Generally it's better to get the conditional logic done at a higher layer. So you have something that looks like this:

```
public void method() {
  if(a || b) {
        doFoo();
         if(a1) {
               doBar()
          }
  } else if(c) {
      doZed();
  }
}
```

vs:

```
public void method() {
   doFoo();
   doBar();
   doZed();
}

private void doFoo() {    
    if(!a && !b && c) {
        doZed();
    } else if(!a && !b) {
        return;
    }
    doBar();
}

private void doBar() {
    if(!a1) {
        return;
    }
}
```

Notice that above the control flow is pushed down and it's very hard to follow what the logic is.

The idea is 3 part:
1. Avoid lists of void method calls
2. Choose where the control flow will be, have the method with control flow just do mainly that and the things it invokes should be relatively functional.
3. Push control flow to higher layers, avoid doing control flow in private methods That indicates you perhaps want a strategy object to handle the behavior.

Reply:
I'm not insisting any changes be made here, but I don't agree that having control flow at a lower layer like this makes it easier. A reader of the code needs to jump to this method anyways and it's cognitively complex to keep track of control flow throughout  'n' layers of code.

Reply:
Expouding on "needs to jump to this method anyways"

Let's say the reader knows that only "a" is true. In this case the following method:

```
public void method() {
  if(a || b) {
        doFoo();
         if(a1) {
               doBar()
          }
  } else if(c) {
      doZed();
  }
}
```

Requires a reader to only read `doFoo`. On the other hand, with the series of void call, a reader needs to go through each method, hence why I suggest often they may as well be inlined, as you have to read it all anyways. The fact you can't skip code (as in this example where you can skip 'doBar' and 'doZed'), is a break-down of information hiding. My 2 cents.

Reply:
I think I understand what you are saying but this entire method is control flow.  So it is like:
```
public void method() {
  if (isAllowed()) {
    doFoo();
  }
}

private boolean isAllowed() {
  return a || (b && c) || d;
}
```
vs
```
public void method() {
  if (a || (b && c) || d) {
    doFoo();
  }
}
```

You just recommended that I do the former in `getNames()` (with the the new `isRetreatPossible()` method).  How is `canAttackerRetreatPartialAmphib()` any different?

Or is it because both `canAttackerRetreatPartialAmphib()` and `canAttackerRetreatAmphibPlanes()` checks for `Properties.getPartialAmphibiousRetreat` and so you want to move that call out of them?